src/core/SafeOp.sol

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.20;
+
+import {
+    PackedUserOperation,
+    UserOperationLib
+} from "@ERC4337/account-abstraction/contracts/core/UserOperationLib.sol";
+import { SAFE_OP_TYPEHASH, ISafeOp } from "../interfaces/ISafeOp.sol";
+
+abstract contract SafeOp is ISafeOp {
+    using UserOperationLib for PackedUserOperation;
+
+    bytes32 private constant DOMAIN_SEPARATOR_TYPEHASH =
+        0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218;
+
+    /**
+     * @dev Decodes an ERC-4337 user operation into a Safe operation.
+     * @param userOp The ERC-4337 user operation.
+     * @return operationData Encoded EIP-712 Safe operation data bytes used for signature
+     * verification.
+     * @return validAfter The timestamp the user operation is valid from.
+     * @return validUntil The timestamp the user operation is valid until.
+     * @return signatures The Safe owner signatures extracted from the user operation.
+     */
+    function _getSafeOp(
+        PackedUserOperation calldata userOp,
+        address entryPoint
+    )
+        internal
+        view
+        returns (
+            bytes memory operationData,
+            uint48 validAfter,
+            uint48 validUntil,
+            bytes calldata signatures
+        )
+    {
+        // Extract additional Safe operation fields from the user operation signature which is
+        // encoded as:
+        // `abi.encodePacked(validAfter, validUntil, signatures)`
+        {
+            bytes calldata sig = userOp.signature;
+            validAfter = uint48(bytes6(sig[0:6]));
+            validUntil = uint48(bytes6(sig[6:12]));
+            signatures = sig[12:];
+        }
+
+        // It is important that **all** user operation fields are represented in the `SafeOp` data
+        // somehow, to prevent
+        // user operations from being submitted that do not fully respect the user preferences. The
+        // only exception is
+        // the `signature` bytes. Note that even `initCode` needs to be represented in the operation
+        // data, otherwise
+        // it can be replaced with a more expensive initialization that would charge the user
+        // additional fees.
+        {
+            // In order to work around Solidity "stack too deep" errors related to too many stack
+            // variables, manually
+            // encode the `SafeOp` fields into a memory `struct` for computing the EIP-712
+            // struct-hash. This works
+            // because the `EncodedSafeOpStruct` struct has no "dynamic" fields so its memory layout
+            // is identical to the
+            // result of `abi.encode`-ing the individual fields.
+            EncodedSafeOpStruct memory encodedSafeOp = EncodedSafeOpStruct({
```

Comment:
this is fixing discussed struct change:
https://github.com/rhinestonewtf/safe7579/pull/3